rfcs/build-system/02-solution-style-ts-configs.md

```diff
@@ -0,0 +1,204 @@
+
+---
+
+_List contributors to the proposal here: @hotell_
+
+## Summary
+
+> **💡 NOTE:**
+>
+> This is proposed changed only for vNext and packages that use new DX ( via ts config path aliases)
+
+We successfully migrated vNext packages [to new DX (1 stage)](https://github.com/microsoft/fluentui/issues/18579), which leverages TS path aliases for improved DX and other benefits...
+
+To continue with our journey to improved DX we need to address leaking global types and other related issues (caused by our current `tsconfig.json` config) for vNext packages.
+
+## Problem statement
+
+> see [Open Issues](#open_issues)
+
+- leaking global types
+- invalid type checking (TS is not following proper `target` and `lib`)
+- various contexts/scopes addressed via 1 config file
+- shipping files besides implementation to npm, thus exposing "private" APIs
+
+## Detailed Design or Proposal
+
+Using [TS solution style config](https://devblogs.microsoft.com/typescript/announcing-typescript-3-9/#solution-style-tsconfig) / per environment configuration, for vNext/packages that use new DX.
+
+### Pre-requirements
+
+- vNext packages migrated to new DX (1st stage) https://github.com/microsoft/fluentui/issues/18579
+
+### Detailed design/Implementation
+
+- ![Configuration and file types processing](https://user-images.githubusercontent.com/1223799/126519445-8c215488-e568-4f88-b37d-b99b4d80f418.png)
+- ![CI build execution](https://user-images.githubusercontent.com/1223799/126519538-c0794d3b-1199-4e49-8ca8-36deeb09be57.png)
+
+#### `<package-name>/tsconfig.json` - solution config file
+
+- main package TS config consumed by TSC Language Server
+- on CI we gonna trigger `nx run <project-name>:type-check` (already happening for `/test`), which will type check properly everything in your package.
+  - under the hood we'll run TS in build mode: `tsc --build`
+
+**Example:**
+
+```json
+{
+  "extends": "../tsconfig.base.json",
+  "compilerOptions": {
+    // by default we gonna use tsc for type checking only
+    "noEmit": true
+    //...
+  },
+  "include": [],
+  "files": [],
+  "references": [
+    {
+      "path": "./tsconfig.lib.json"
+    },
+    {
+      "path": "./tsconfig.spec.json"
+    },
+    {
+      "path": "./storybook/tsconfig.json"
+    }
+  ]
+}
+```
+
+#### `<package-name>/tsconfig.lib.json`
+
+- config used for production builds
+
+**Example:**
+
+```json
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "ES2015",
+    "outDir": "../dist/out-tsc",
+    // properly scoped globals and environment
+    "types": [],
+    // enable transpilation and declaration files generation only for implementation files
+    "noEmit": false,
+    "declaration": true
+  },
+  "exclude": ["**/*.spec.ts", "**/*.test.ts", "**/*.stories.tsx"],
+  "include": ["**/*.ts", "**/*.tsx"]
+}
+```
+
+#### `<package-name>/tsconfig.spec.json`
+
+- config used for tests
+
+**Example:**
+
+```json
+{
+  "extends": "./tsconfig.json",
```

Comment:
Could be a stupid question, are there any circular issues since each tsconfig is extending the root `tsconfig.json` which is then also referencing each tsconfig 🤔🤔